srtcore/core.cpp

```diff
@@ -9193,7 +9193,7 @@ int srt::CUDT::packLostData(CPacket& w_packet)
 
         // The packet has been ecrypted, thus the authentication tag is expected to be stored
         // in the SND buffer as well right after the payload.
-        if (m_config.iCryptoMode == CSrtConfig::CIPHER_MODE_AES_GCM)
+        if (m_pCryptoControl->getCryptoMode() == CSrtConfig::CIPHER_MODE_AES_GCM)
```

Comment:
Check if crypto control exists. Also `m_ConnectionLock` needs to be locked, and it is locked in `packData(..)` up the stack, so all good there.
```suggestion
        if (m_pCryptoControl && m_pCryptoControl->getCryptoMode() == CSrtConfig::CIPHER_MODE_AES_GCM)
```

Reply:
Done!

Reply:
Also, is this block of code susceptible to the same issue too? It seems to be doing some initialisation at the start, does it matter if the mode is auto at this stage?

```
    try
    {
        const int authtag = m_config.iCryptoMode == CSrtConfig::CIPHER_MODE_AES_GCM ? HAICRYPT_AUTHTAG_MAX : 0;
        m_pSndBuffer = new CSndBuffer(32, m_iMaxSRTPayloadSize, authtag);
        SRT_ASSERT(m_iISN != -1);
        m_pRcvBuffer = new srt::CRcvBuffer(m_iISN, m_config.iRcvBufSize, m_pRcvQueue->m_pUnitQueue, m_config.bMessageAPI);
        // after introducing lite ACK, the sndlosslist may not be cleared in time, so it requires twice space.
        m_pSndLossList = new CSndLossList(m_iFlowWindowSize * 2);
        m_pRcvLossList = new CRcvLossList(m_config.iFlightFlagSize);
    }
 ```

This is inside srt::CUDT::prepareConnectionObjects.

Reply:
Yes. The difficulty here is that CryptoControl is not yet created. Probably worth checking against `CIPHER_MODE_AES_CTR` and adding auth tag both for CIPHER_MODE_AES_AUTO and CIPHER_MODE_AES_GCM. 🤔

Reply:
Or create CryptoControl first... Let me check.

Reply:
Moving `CryptoControl` up won't help. For the responder, the mode is deduced later when processing the Key Material (KM) request in `interpretSrtHandshake`.
```c++
        ok = ok &&  prepareConnectionObjects(m_ConnRes, m_SrtHsSide, eout);

        // May happen that 'response' contains a data packet that was sent in rendezvous mode.
        // In this situation the interpretation of handshake was already done earlier.
        ok = ok && pResponse->isControl();
        ok = ok && interpretSrtHandshake(m_ConnRes, *pResponse, 0, 0);
```

Reply:
Right. So is there any downside to just doing as you suggested...

```
Probably worth checking against CIPHER_MODE_AES_CTR and adding auth tag both for CIPHER_MODE_AES_AUTO and CIPHER_MODE_AES_GCM
```

...apart from having an extra 16 bytes of unnecessary buffer space in CTR mode?

Reply:
The downside would be 16 bytes less space for the payload in Auto mode: 1440 instead of 1456.

Reply:
Ok, so it just reduces the available space, but it's not like something is going to overwrite memory or something?

Reply:
I think the payload might take two packets instead of one in that case, which is unwanted in live mode.

Reply:
Ah, no, sorry. This has been fixed in #2589.

Reply:
Ok, I've committed a change to reserve the space in both AUTO and GCM modes, for now.

Maybe your proper patch can follow later? Well, up to you guys obviously.